File: src/main/java/com/milaboratory/mixcr/trees/Tree.java
/*
 * Copyright (c) 2014-2019, Bolotin Dmitry, Chudakov Dmitry, Shugay Mikhail
 * (here and after addressed as Inventors)
 * All Rights Reserved
 *
 * Permission to use, copy, modify and distribute any part of this program for
 * educational, research and non-profit purposes, by non-profit institutions
 * only, without fee, and without a written agreement is hereby granted,
 * provided that the above copyright notice, this paragraph and the following
 * three paragraphs appear in all copies.
 *
 * Those desiring to incorporate this work into commercial products or use for
 * commercial purposes should contact MiLaboratory LLC, which owns exclusive
 * rights for distribution of this program for commercial purposes, using the
 * following email address: licensing@milaboratory.com.
 *
 * IN NO EVENT SHALL THE INVENTORS BE LIABLE TO ANY PARTY FOR DIRECT, INDIRECT,
 * SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING LOST PROFITS,
 * ARISING OUT OF THE USE OF THIS SOFTWARE, EVEN IF THE INVENTORS HAS BEEN
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * THE SOFTWARE PROVIDED HEREIN IS ON AN "AS IS" BASIS, AND THE INVENTORS HAS
 * NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR
 * MODIFICATIONS. THE INVENTORS MAKES NO REPRESENTATIONS AND EXTENDS NO
 * WARRANTIES OF ANY KIND, EITHER IMPLIED OR EXPRESS, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A
 * PARTICULAR PURPOSE, OR THAT THE USE OF THE SOFTWARE WILL NOT INFRINGE ANY
 * PATENT, TRADEMARK OR OTHER RIGHTS.
 */
package com.milaboratory.mixcr.trees;

import javax.annotation.Nullable;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Stream;

/**
 *
 */
public class Tree<T> {
    private final Node<T> root;

    public Tree(Node<T> root) {
        this.root = root;
    }

    public Node<T> getRoot() {
        return root;
    }

    public Stream<NodeWithParent<T>> allNodes() {
        return Stream.concat(
                Stream.of(new NodeWithParent<>(null, root, null)),
                root.allDescendants()
        );
    }

    public <R> Tree<R> map(Function<T, R> mapper) {
        return new Tree<>(root.map(mapper));
    }

    public static class Node<T> {
        private final T content;
        private final List<NodeLink<T>> children;

        protected Node(T content) {
            this.content = content;
            children = new ArrayList<>();
        }

        protected Node(T content, List<NodeLink<T>> children) {
            this.content = content;
            this.children = children;
        }

        public List<NodeLink<T>> getLinks() {
            return children;
        }

        public Node<T> addChild(Node<T> node) {
            return addChild(node, null);
        }

        public Node<T> addChild(Node<T> node, @Nullable BigDecimal distance) {
            children.add(new NodeLink<>(node, distance));
            return this;
        }

        public T getContent() {
            return content;
        }

        public void replaceChild(Node<T> what, Node<T> substitution, @Nullable BigDecimal distance) {
            if (!children.removeIf(it -> it.node == what)) {
                throw new IllegalArgumentException();
            }
            children.add(new NodeLink<>(substitution, distance));
        }

        public Stream<NodeWithParent<T>> allDescendants() {
            return children.stream()
                    .flatMap(link -> Stream.concat(
                            Stream.of(new NodeWithParent<>(this, link.getNode(), link.getDistance())),
                            link.getNode().allDescendants())
                    );
        }

        public <R> Node<R> map(Function<T, R> mapper) {
            Node<R> mappedNode = new Node<>(mapper.apply(content));
            children.forEach(child -> mappedNode.addChild(child.node.map(mapper), child.distance));
            return mappedNode;
        }
    }

    static class NodeLink<T> {
        private final Node<T> node;
        @Nullable
        private final BigDecimal distance;

        public NodeLink(Node<T> node, @Nullable BigDecimal distance) {
            this.node = node;
            this.distance = distance;
        }

        public Node<T> getNode() {
            return node;
        }

        @Nullable
        public BigDecimal getDistance() {
            return distance;
        }
    }

    public static class NodeWithParent<T> {
        @Nullable
        private final Node<T> parent;
        private final Node<T> node;
        @Nullable
        private final BigDecimal distance;

        public NodeWithParent(@Nullable Node<T> parent, Node<T> node, @Nullable BigDecimal distance) {
            this.parent = parent;
            this.node = node;
            this.distance = distance;
        }

        @Nullable
        public BigDecimal getDistance() {
            return distance;
        }

        @Nullable
        public Node<T> getParent() {
            return parent;
        }

        public Node<T> getNode() {
            return node;
        }
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/trees/Tree.kt
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.trees

import com.milaboratory.primitivio.PrimitivI
import com.milaboratory.primitivio.PrimitivO

/**
 *
 */
class Tree<T : Any>(
    val root: Node<T>
) {
    fun copy(): Tree<T> = Tree(root.copy())

    fun allNodes(): List<NodeWithParent<T>> {
        val result = mutableListOf<NodeWithParent<T>>()
        result += NodeWithParent(null, root, null)
        root.registerAllDescendants(result)
        return result
    }

    /** List of all leafs in the t@ree */
    fun allLeafs() = allNodes().asSequence().filter { it.node.isLeaf() }

    fun <R : Any> map(mapper: (T?, T) -> R): Tree<R> =
        Tree(root.map(null, mapper))

    class Node<T> {
        val content: T
        private val children: MutableList<NodeLink<T>>

        constructor(content: T) {
            this.content = content
            children = mutableListOf()
        }

        constructor(content: T, children: List<NodeLink<T>>) {
            this.content = content
            this.children = children.toMutableList()
        }

        fun copy(): Node<T> {
            val result = Node(content)
            children.forEach { result.addChild(it.node.copy(), it.distance) }
            return result
        }

        /** Returns the height of the tree */
        fun height(): Int = if (children.isEmpty()) 0 else 1 + children.maxOf { it.node.height() }

        val links: List<NodeLink<T>>
            get() = children

        fun addChild(node: Node<T>, distance: Double): Node<T> {
            children.add(NodeLink(node, distance))
            return this
        }

        fun replaceChild(what: Node<T>, substitution: Node<T>, distance: Double) {
            require(children.removeIf { it.node === what })
            addChild(substitution, distance)
        }

        fun registerAllDescendants(result: MutableList<NodeWithParent<T>>) {
            children.forEach { link ->
                result += NodeWithParent(this, link.node, link.distance)
                link.node.registerAllDescendants(result)
            }
        }

        fun <R> map(parentContent: T?, mapper: (T?, T) -> R): Node<R> {
            val result = Node(mapper(parentContent, content))
            children.forEach { child ->
                result.addChild(
                    child.node.map(content, mapper),
                    child.distance
                )
            }
            return result
        }

        /** Whether this node is leaf */
        fun isLeaf() = links.isEmpty()
    }

    class NodeLink<T>(
        val node: Node<T>,
        val distance: Double
    )

    data class NodeWithParent<T>(
        val parent: Node<T>?,
        val node: Node<T>,
        val distance: Double?
    )
}

object TreeSerializer {
    fun writeTree(output: PrimitivO, obj: Tree<*>) {
        writeNode(output, obj.root)
    }

    private fun writeNode(output: PrimitivO, node: Tree.Node<*>) {
        output.writeObject(node.content)
        output.writeInt(node.links.size)
        node.links.forEach { link ->
            output.writeDouble(link.distance)
            writeNode(output, link.node)
        }
    }

    inline fun <reified T : Any> readTree(input: PrimitivI): Tree<T> {
        return Tree(readNode(input, T::class.java))
    }

    fun <T> readNode(input: PrimitivI, klass: Class<T>): Tree.Node<T> {
        val content = input.readObject(klass)
        val count = input.readInt()
        val result = Tree.Node(content)
        repeat(count) {
            val distance = input.readDouble()
            val child = readNode(input, klass)
            result.addChild(child, distance)
        }
        return result
    }
}
